src/ConnectedMode/ServerSentEvents/SSESessionFactory.cs

```diff
@@ -0,0 +1,185 @@
+            this.sonarQubeClient = sonarQubeClient;
+            this.issueChangedServerEventSourcePublisher = issueChangedServerEventSourcePublisher;
+            this.taintServerEventSourcePublisher = taintServerEventSourcePublisher;
+            this.threadHandling = threadHandling;
+        }
+
+        public ISSESession Create(string projectKey)
+        {
+            if (disposed)
+            {
+                throw new ObjectDisposedException(nameof(SSESessionFactory));
+            }
+
+            var session = new SSESession(issueChangedServerEventSourcePublisher,
+                taintServerEventSourcePublisher,
+                projectKey,
+                threadHandling,
+                sonarQubeClient);
+            
+            return session;
+        }
+
+        public void Dispose()
+        {
+            if (disposed)
+            {
+                return;
+            }
+
+            issueChangedServerEventSourcePublisher.Dispose();
+            taintServerEventSourcePublisher.Dispose();
+            disposed = true;
+        }
+
+        internal sealed class SSESession : ISSESession
+        {
+            private readonly IIssueChangedServerEventSourcePublisher issueChangedServerEventSourcePublisher;
+            private readonly ITaintServerEventSourcePublisher taintServerEventSourcePublisher;
+            private readonly string projectKey;
+            private readonly IThreadHandling threadHandling;
+            private readonly ISonarQubeService sonarQubeService;
+            private readonly CancellationTokenSource sessionTokenSource;
+
+            private bool disposed;
+
+            internal SSESession(IIssueChangedServerEventSourcePublisher issueChangedServerEventSourcePublisher, 
+                ITaintServerEventSourcePublisher taintServerEventSourcePublisher,
+                string projectKey,
+                IThreadHandling threadHandling,
+                ISonarQubeService sonarQubeService)
+            {
+                this.issueChangedServerEventSourcePublisher = issueChangedServerEventSourcePublisher;
+                this.taintServerEventSourcePublisher = taintServerEventSourcePublisher;
+                this.projectKey = projectKey;
+                this.threadHandling = threadHandling;
+                this.sonarQubeService = sonarQubeService;
+                this.sessionTokenSource = new CancellationTokenSource();
+            }
+
+            public async Task PumpAllAsync()
+            {
+                if (disposed)
+                {
+                    throw new ObjectDisposedException(nameof(SSESession));
+                }
+
+                await threadHandling.SwitchToBackgroundThread();
+
+                var sseStream = await sonarQubeService.CreateServerSentEventsStream(projectKey, sessionTokenSource.Token);
+
+                if (sseStream == null)
+                {
+                    return;
+                }
+
+                sseStream.BeginListening().Forget();
```

Comment:
FYI I added this following the merge conflict. I _think_ we might need to run it on another thread. I'm not sure. It uses the underlying network ReadLineAsync, and I _think_ it shouldn't block (or it won't if we use ReadLineAsync().ConfigureAwait(false)).